kcidb/test_oo.py

```diff
@@ -203,3 +203,240 @@ def test_two_object(self):
             '>checkout#>build#>test#',
             {"test", "build", "checkout", "revision"}, 2
         )
+
+
+class KCIDBTraversingTestCase(kcidb.unittest.TestCase):
+    """Test case for traversing objects."""
+
+    # Conform to unittest conventions, pylint: disable=invalid-name
+
+    def query_str(self, pattern_string):
+        """Run OO client query with a parsed string."""
+        return self.client.query(
+            kcidb.orm.Pattern.parse(pattern_string)
+        )
+
+    def assertListContains(self, list_of_items, num_elements, type_name):
+        """
+        Check that the list of item contains the correct number of
+        element(s) and that the element(s) are the same type.
+        Args:
+            list_of_items:      A list of items.
+            num_elements:       The number of element(s) contained
+                                in the list of item.
+            type_name:          The expected schema type name.
+        """
+        assert isinstance(list_of_items, list)
+        assert isinstance(num_elements, int)
+        assert isinstance(type_name, str)
+
+        self.assertEqual(
+            len(list_of_items), num_elements,
+            msg=f"Invalid number of {type_name} instances"
+        )
+        for i, obj in enumerate(list_of_items):
+            self.assertIsInstance(obj, kcidb.oo.CLASSES[type_name],
+                                  msg=f"Invalid type of {type_name} #{i}")
+
+    def assertDictContains(self, dict_of_items, data, type_name):
```

Comment:
I think we can get away with just one function, for both kinds of containers. E.g.:

```python
def assertContains(self, container, contents):
    """
    Check that a container has the specified contents.

    Args:
        container:  The container to check the contents of.
        contents:   The contents pattern to check against. The pattern matches
                    either a list of certain number of objects of particular
                    type, or a dictionary with string keys and values matching
                    specified patterns.

                    A list pattern is a tuple containing the type of each
                    contained instance, and the number of those instances.

                    A dictionary pattern is a dictionary containing string
                    keys for the matched dictionary to have, with values being
                    patterns for the corresponding values in the matched
                    dictionary.

    Returns:
        True if the container matched the contents pattern, False otherwise.
    """
```

If we make the patterns contain types directly, instead of their names in kcidb.oo.CLASSES, then the invocations will be more direct and obvious, and we can make the function work with more types.

You can, of course, import the types you need directly, so you can e.g. write `Checkout` instead of `kcidb.oo.Checkout`, and make the code shorter.

BTW, we also need to check that dictionaries don't contain keys we did not specify.

Reply:
> You can, of course, import the types you need directly, so you can 

Importing directly works perfectly with `Checkout` and `Build`, but gives warning with `Test` because **unittest** tries to collect the Test class but it couldn't because it has a ` __init__ ` constructor. 

I added `__test__= False` to prevent `Test` from being collected by **unittest**.